prime-router/src/main/kotlin/azure/ActionHistory.kt

```diff
@@ -713,4 +732,166 @@ class ActionHistory {
+                    it.writeStringField("details", error.responseMessage.detailMsg())
+                    it.writeEndObject()
+                }
+                it.writeEndArray()
+            }
+            writeDetailsArray("errors", errors)
+            writeDetailsArray("warnings", warnings)
+
+            fun createRowsDescription(rows: List<Int>?): String {
+                // Consolidate row ranges, e.g. 1,2,3,5,7,8,9 -> 1-3,5,7-9
+                if (rows == null || rows.isEmpty()) return ""
+                val sb = StringBuilder().append("Rows: ")
+                var isListing = false
+                rows.sorted().forEachIndexed { i, row ->
+                    if (i == 0) {
+                        sb.append(row.toString())
+                    } else if (row == rows[i - 1] || row == rows[i - 1] + 1) {
+                        isListing = true
+                    } else if (isListing) {
+                        sb.append(" to ${rows[i - 1]}, $row")
+                        isListing = false
+                    } else {
+                        sb.append(", $row")
+                        isListing = false
+                    }
+                    if (i == rows.lastIndex && isListing) {
+                        sb.append(" to ${rows[rows.lastIndex]}")
+                    }
+                }
+                return sb.toString()
+            }
+
+            fun writeConsolidatedArray(field: String, array: List<ResultDetail>) {
+                val rowsByGroupingId = hashMapOf<String, MutableList<Int>>()
+                val messageByGroupingId = hashMapOf<String, String>()
+                array.forEach { resultDetail ->
+                    val groupingId = resultDetail.responseMessage.groupingId()
+                    if (!rowsByGroupingId.containsKey(groupingId)) {
+                        rowsByGroupingId[groupingId] = mutableListOf()
+                        messageByGroupingId[groupingId] = resultDetail.responseMessage.detailMsg()
+                    }
+                    if (resultDetail.row != -1) {
+                        // Add 2 to account for array offset and csv header
+                        rowsByGroupingId[groupingId]?.add(resultDetail.row + 2)
+                    }
+                }
+                it.writeArrayFieldStart(field)
+                rowsByGroupingId.keys.forEach { groupingId ->
+                    it.writeStartObject()
+                    it.writeStringField("message", messageByGroupingId[groupingId])
+                    it.writeStringField("rows", createRowsDescription(rowsByGroupingId[groupingId]))
+                    it.writeEndObject()
+                }
+                it.writeEndArray()
+            }
+            writeConsolidatedArray("consolidatedErrors", errors)
+            writeConsolidatedArray("consolidatedWarnings", warnings)
+        }
+        return outStream.toString()
+    }
+
+    /**
+     * Creates a list of [ItemRouting] instances with the report index
```

Comment:
Do not use the [] in the [ItemRouting] as that is not an argument to this function.